### PR TITLE
Fix WithSpan initialization ordering requirement for SpanAttribute and Flux

### PR DESCRIPTION
**Pull request by anuraaga**
Also extract shared tests for WithSpan and apply to Guava and reactor

### CODE REVIEW

**Comment by anuraaga** [comment]
All I wanted was to extract tests and didn't expect the rabbit hole :-O

File: instrumentation/opentelemetry-annotations-1.0/testing/src/main/java/io/opentelemetry/javaagent/instrumentation/otelannotations/AbstractTraced.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+public abstract class AbstractTraced<T extends U, U> {
+
+  protected static final String SUCCESS_VALUE = "Value";
+
+  protected static final IllegalArgumentException FAILURE = new IllegalArgumentException("Boom");
+
+  protected AbstractTraced() {
+    if (!getClass().getSimpleName().equals("Traced")) {
+      throw new IllegalStateException("Subclasses of AbstractTraced must be named Traced");
```
Comment (anuraaga) [inline instrumentation/opentelemetry-annotations-1.0/testing/src/main/java/io/opentelemetry/javaagent/instrumentation/otelannotations/AbstractTraced.java:16]: I could instead use `getClass` in the assertions themselves but this made the assertions much nicer

File: instrumentation/methods/testing/src/main/java/io/opentelemetry/javaagent/instrumentation/methods/AbstractWithTraceSpanTest.java [hunk 2]
```diff
@@ -1,20 +1,15 @@
...
+package io.opentelemetry.javaagent.instrumentation.methods;
 
 import com.google.common.util.concurrent.Futures;
 import com.google.common.util.concurrent.ListenableFuture;
 import com.google.common.util.concurrent.SettableFuture;
-import io.opentelemetry.extension.annotations.WithSpan;
 
-final class TracedWithSpan {
-  static final IllegalArgumentException FAILURE = new IllegalArgumentException("Boom");
+public abstract class AbstractWithTraceSpanTest<T, U> {
```
Comment (mateuszrzeszutek) [inline instrumentation/methods/testing/src/main/java/io/opentelemetry/javaagent/instrumentation/methods/AbstractWithTraceSpanTest.java:7]: Is this classed used anywhere?

File: instrumentation/methods/testing/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+plugins {
```
Comment (mateuszrzeszutek) [inline instrumentation/methods/testing/build.gradle.kts:1]: This module does not seem to be used

File: instrumentation/reactor/reactor-3.1/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/reactor/HooksInstrumentation.java [hunk 4]
```diff
@@ -19,7 +19,9 @@
 public class HooksInstrumentation implements TypeInstrumentation {
   @Override
   public ElementMatcher<TypeDescription> typeMatcher() {
-    return named("reactor.core.publisher.Hooks");
+    return named("reactor.core.publisher.Hooks")
+        // Hooks may not be loaded early enough so also match our main targets
+        .or(named("reactor.core.publisher.Flux").or(named("reactor.core.publisher.Mono")));
```
Comment (mateuszrzeszutek) [inline instrumentation/reactor/reactor-3.1/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/reactor/HooksInstrumentation.java:24]: Nit: you can use `namedOneOf` instead

Comment (mateuszrzeszutek) [inline instrumentation/reactor/reactor-3.1/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/reactor/HooksInstrumentation.java:24]: Do `Mono` and `Flux` even have a static initializer method?

File: instrumentation/methods/testing/build.gradle.kts (hunk 3 above)
Comment (anuraaga) [reply instrumentation/methods/testing/build.gradle.kts]: Oops sorry

File: instrumentation/reactor/reactor-3.1/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/reactor/HooksInstrumentation.java (hunk 4 above)
Comment (anuraaga) [reply instrumentation/reactor/reactor-3.1/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/reactor/HooksInstrumentation.java:24]: Yeah they declare some constants. The tests fail without adding these

**Review by mateuszrzeszutek — APPROVED** [review]

**Pull request merged**